### ISSUE TITLE
Elements become entangled when using ternary operator

### ISSUE THREAD

**Issue by q3452** [react-toolbox/react-toolbox#1156]
In React, one form of dynamic element creation is the use of the ternary operator to swap elements based on document state.

In React-toolbox, this seems to cause elements in the same DOM position to become entangled in some way that causes a null value in one element to be overriden with a previous value in another element.

This code snippet below will demonstrate the problem when used in the React Toolbox sandbox.  Simple click the toggle button and observe the first element's value appearing in the second element.  Any value entered for the second element will correct the bug but it exists until a value is entered.

```
class NullValueBug extends React.Component {
  state = { firstPanel: true, text1: 'This should only appear in panel 1', text2: null };

  handleChange = (name, value) => {
    this.setState({...this.state, [name]: value});
  };
  
  togglePanels = () => {
  	this.setState({firstPanel: !this.state.firstPanel});
   }
  render () {
    return (
      <section>
    		<Button label='Switch' accent onClick={this.togglePanels} />
        { (this.state.firstPanel) ?
        <Input type='text' label='text1' name='text1' value={this.state.text1} />
        :
        <Input type='text' label='text2' name='text2' value={this.state.text2} onChange={this.handleChange.bind(this, 'text2')} />
        }
      </section>
    );
  }
}

return <NullValueBug />;

```

**Comment by javivelasco** [comment]
I don't see this as an issue. The problem here is that you are passing a `null` value to the `Input` component value. You get this warning in the console:

```
Warning: `value` prop on `input` should not be null. Consider using the empty string to clear the component or `undefined` for uncontrolled components.
```

If you change `null` to an empty string it works 👌

**Comment by q3452** [comment]
An empty string and a null value are not the same thing.  From a DB perspective, a null is an unset value and an empty string is a value explicitly set to a blank string.

Being able to set a value as null is important to avoid overwriting an unset value with a 0 length string, which implies a choice by the user.

Consider a page with a read-from-DB, display, write-to-DB loop.  If we set it to an empty string, the written values do not match the read values even if the user hasn't touched them.  That creates side-effects.

From a DB perspective and a software-engineering perspective, these are serious issues.

**Comment by panjiesw** [comment]
The native `<input />` of React itself can't receive a `null` value. Some discussion about `null` values on inputs:

facebook/react#6996
http://stackoverflow.com/questions/17240432/insert-null-value-into-a-text-input
http://stackoverflow.com/questions/36187522/react-select-with-value-null

**Comment by q3452** [comment]
@javivelasco 
I encountered a very similar problem again with Dropdowns which I had wrapped in my own control (it added DB access code) which made me realise that this is also Key related.  React-Toolbox isn't providing a key on its elements which causes React to generate one based on DOM position.  Explicitly setting a key prevents the issue above from occurring.  In the above example, it would look like this:
`<Input type='text' label='text1' name='text1 key='text1' ...`

If the framework _or_ the developer set the key to the name or something, the problem doesn't occur.  It's also specific to dynamic controls since these are the only ones that change position in the DOM.  I can't decide whether this is actually the responsibility of the framework or not but either way, your examples don't do it and it should either be done explicitly by the developer (who follows your examples) or implicitly by the framework when a key isn't provided.
